File: myExpenses/src/debug/java/org/totschnig/myexpenses/provider/TransactionInfo.java
package org.totschnig.myexpenses.provider;

import android.content.ContentValues;

import org.totschnig.myexpenses.model.Model;
import org.totschnig.myexpenses.model.CrStatus;

public class TransactionInfo {

  String comment;
  long amount;
  long date;
  long payeeId;
  long accountId;

  /*
   * Constructor for a TransactionInfo instance. This class helps create a transaction and
   * return its values in a ContentValues map expected by data model methods.
   * The transaction's id is created automatically when it is inserted into the data model.
   */
  public TransactionInfo(String comment, long date, long amount, long accountId, long payeeId) {
    this.comment = comment;
    this.date = date;
    this.amount = amount;
    this.payeeId = payeeId;
    this.accountId = accountId;
  }

  /*
   * Returns a ContentValues instance (a map) for this TransactionInfo instance. This is useful for
   * inserting a TransactionInfo into a database.
   */
  public ContentValues getContentValues() {
    // Gets a new ContentValues object
    ContentValues v = new ContentValues();

    // Adds map entries for the user-controlled fields in the map
    v.put(DatabaseConstants.KEY_COMMENT, comment);
    v.put(DatabaseConstants.KEY_DATE, date);
    v.put(DatabaseConstants.KEY_VALUE_DATE, date);
    v.put(DatabaseConstants.KEY_AMOUNT, amount);
    v.put(DatabaseConstants.KEY_PAYEEID, payeeId);
    v.put(DatabaseConstants.KEY_ACCOUNTID, accountId);
    v.put(DatabaseConstants.KEY_CR_STATUS, CrStatus.UNRECONCILED.name());
    v.put(DatabaseConstants.KEY_UUID, Model.generateUuid());
    return v;
  }


  public String getComment() {
    return comment;
  }

  public long getAmount() {
    return amount;
  }

  public long getDate() {
    return date;
  }

  public long getPayeeId() {
    return payeeId;
  }

  public long getAccountId() {
    return accountId;
  }
}

File: myExpenses/src/debug/java/org/totschnig/myexpenses/provider/TransactionInfo.kt
package org.totschnig.myexpenses.provider

import android.content.ContentValues
import org.totschnig.myexpenses.model.CrStatus
import org.totschnig.myexpenses.model.generateUuid
import org.totschnig.myexpenses.util.toEpoch
import java.time.LocalDateTime

data class TransactionInfo @JvmOverloads constructor(
    val accountId: Long,
    val amount: Long,
    val date: LocalDateTime = LocalDateTime.now(),
    val comment: String = "",
    val payeeId: Long? = null,
    val debtId: Long? = null,
    val catId: Long? = null,
    val methodId: Long? = null,
    val parentId: Long? = null,
    val crStatus: CrStatus = CrStatus.UNRECONCILED,
    val equivalentAmount: Long? = null
) {
    val dateAsLong: Long
        get() = date.toEpoch()
    val contentValues: ContentValues
        get() = ContentValues().apply {
            put(KEY_COMMENT, comment)
            put(KEY_DATE, dateAsLong)
            put(KEY_VALUE_DATE, dateAsLong)
            put(KEY_AMOUNT, amount)
            put(KEY_PAYEEID, payeeId)
            put(KEY_ACCOUNTID, accountId)
            put(KEY_CR_STATUS, crStatus.name)
            put(KEY_UUID, generateUuid())
            put(KEY_DEBT_ID, debtId)
            put(KEY_CATID, catId)
            put(KEY_METHODID, methodId)
            put(KEY_PARENTID, parentId)
            equivalentAmount?.let { put(KEY_EQUIVALENT_AMOUNT, it) }
        }
}
